### PR TITLE
Update yoga and allow building with asm+emscripten

### PR DESCRIPTION
**Pull request by victorporof**
This PR does 2 things:

1. Downgrades bindgen to a version that works when cross-compiling for asm+emscripten. I could not for my life figure out how to make this work with the last version of bindgen, which seems to have some bugs around making bindings that also work for emscripten after already compiling as a static library. The older version doesn't seem to mess up anything.

2. Updates yoga to the latest version and switches to a whitelist approach for symbols (thus inherently making obsolete all of the past issues with FP_* symbols not working and needing to update bindgen), since the latest versions use c++ which necessarily requires whitelisting. I needed this because the latest versions of yoga fix some nasty aspect-ratio related bugs.

Your choice if you want to merge this or not.

### CODE REVIEW

**Comment by victorporof** [comment]
Test changes follow this commit: https://github.com/facebook/yoga/commit/35a9f33abb85f08530df0e6b58eba50311c5e752#diff-807659cd883fc0a63f6ce615893b3558

**Comment by bschwind** [comment]
Wow this is great news, and right after #36 was filed! I'm going to let it finish on Travis and take a look at the changes, but it definitely seems like something we want to do or we'll be left way behind on features in the main repo. I was kind of disappointed to hear they switched to C++ for Yoga but if you've already found success using bindgen with it then that's great!

**Comment by victorporof** [comment]
I've also hit memory corruption issues which seem to go away when not using transmute in this crate. I've added a commit to remove that, which simply requires that callbacks use the bindgen-generated data structures instead of our own versions (which in this particular case are identical). Since From and Into are implemented to convert between them, I think this is far better than relying on transmute. The API abstraction was leaky anyway since we were passing internal::NodeRefs instead Nodes.

**Comment by bschwind** [comment]
100% agree on removing transmute, I wanted to get away from that as quickly as possible. Sorry I haven't had much time to review, but I always appreciate the improvements you make!

**Comment by victorporof** [comment]
No worries, working my way through the travis failures now

**Comment by victorporof** [comment]
Travis is only failing on Linux. It seems that the compiler there simply ignores the `-std=c++11` flag when building, which leads me to assume that something is wrong with the compiler itself since the flag is indeed properly passed in the generated build command.

Any help investigating this would be appreciated, since I have a few other things to focus on at the moment. It seems like a potentially easy thing to fix once we figure out why that flag is ignored by the compiler, since the problem is simply that features in the c++11 revision aren't supported/used.

**Comment by bschwind** [comment]
Gotcha, I'll take a look!

**Comment by victorporof** [comment]
Thanks @bschwind! Take your time, I'm perfectly happy using this in a branch for near future.

**Comment by bschwind** [comment]
My internet is being terrible at the moment so you might see a lot of test builds running for trivial changes :P

**Comment by victorporof** [comment]
Fire a blaze

**Comment by bschwind** [comment]
Seems like I can get Linux to work by forcing it to use clang, but now on OS X is failing to find basic includes like `#include <array>`. This seems to be down to the `--target=x86_64-apple-darwin`, because I can omit it and get a successful build. It only adds that flag if the C++ compiler is "clang-like", so letting it default to the `c++` compiler (which is actually clang++ anyway) will sort of trick it into not adding that flag...I guess.

I'm running some more tests on travis to see what these values are for linux as well. C++ build environments are probably one of the most depressing aspects of software development.

File: src/lib.rs [hunk 1]
```diff
@@ -13,6 +13,7 @@ extern crate serde_derive;
 mod internal {
 	#![allow(dead_code)]
 	include!(concat!(env!("OUT_DIR"), "/bindings.rs"));
+	pub use self::root::*;
```
Comment (bschwind) [inline src/lib.rs]: I'm not particularly great with Rust modules, can you tell me what this is doing exactly? I'm not sure how a `pub use` works inside of a private module, and am not entirely sure what module `self::root` refers to.

**Comment by bschwind** [comment]
@victorporof looks good! I just have a small question about modules.

As for the Linux build situation, I have a pull request into your branch here:

https://github.com/victorporof/yoga-rs/pull/1

Feel free to just integrate that change into the PR here if you prefer. Once that's addressed I think this is good to merge.

**Comment by victorporof** [comment]
Apologies this takes longer than expected. Things are looking good but I'm preparing a few additional changes that I'd like to merge (namely exporting a few more yoga functions that we missed). I'll send you over some updates asap.

File: src/lib.rs (hunk 1 above)
Comment (victorporof) [reply src/lib.rs]: This is due to the bindgen changes. Building with c++ namespaces puts all externals inside a module called root.

Comment (bschwind) [reply src/lib.rs]: Ahh okay, thanks for the explanation!

**Comment by bschwind** [comment]
@victorporof how is this looking now? There's another PR, #38, which updates to bindgen `0.36.0`, perhaps that has fixed your issue with asm+emscripten. I'll try to check it out later.

I'd like to get this merged fairly soon since it's such a large change. I have a [PR](https://github.com/victorporof/yoga-rs/pull/1) open in your fork of yoga to fix the travis build, though it's a bit hacky. Maybe I can revisit it and come up with something better...

**Comment by victorporof** [comment]
Indeed, I’m not happy this is taking so long. Feel free to merge, I’m happy with the working (although hacky) changes.

**Comment by cztomsik** [comment]
hey guys, what's the status on this? I'd like to upgrade yoga to latest version but maybe we want to merge this first?

**Comment by bschwind** [comment]
Hey @cztomsik , sorry this has lingered for so long. It's not in an active status, and it seems like it needs quite a bit of work to get it updated with the latest code. I could give it some time in the next week two but I currently have a lot of time to devote to it.

**Comment by cztomsik** [comment]
> Hey @cztomsik , sorry this has lingered for so long. It's not in an active status, and it seems like it needs quite a bit of work to get it updated with the latest code. I could give it some time in the next week two but I currently have a lot of time to devote to it.

hey, no worries, I am not using yoga anymore.

**Pull request closed without merging**